### PR TITLE
Added Xml Encryption capabilities.

### PR DESCRIPTION
**Pull request by kennetlindberg**
This adds the ability to encrypt/decrypt Xml Documents by wrapping the EncryptedXml DotNet classes.

### CODE REVIEW

**Review by PredragMaricic — CHANGES_REQUESTED** [review]

**Review by brentholtorf — APPROVED** [review]

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        //Convert XmlDocument to DotNet XmlDocument preserving whitespace
+        XmlDotNetConvert.ToDotNet(XmlDocument, DotNetXmlDocument, true);
+
+        //Get the DotNet XmlElement to encrypt
+        DotNetXmlElementToEncrypt := DotNetXmlDocument.GetElementsByTagName(ElementToEncrypt).Item(0);
+        if IsNull(DotNetXmlElementToEncrypt) then
+            Error(ElementNotFoundErr, ElementToEncrypt);
+
+        //Initialize a X509Certificate2.
+        X509CertificateImpl.InitializeX509Certificate(X509CertBase64Value, '', DotNetX509Certificate2);
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al:38]: Well, I actually changed it to _internal_ so it works, but is that not ok?

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+        //Create a EncryptedData xml element
+        DotNetEncryptedData := DotNetEncryptedData.EncryptedData();
+        DotNetEncryptedData.CipherData().CipherValue := DotNetEncryptedDataBytes;
+        DotNetEncryptedData.Type := XmlEncElementUrlTok;
+        DotNetEncryptedData.EncryptionMethod :=
+            DotNetEncryptionMethod.EncryptionMethod(SymmetricAlgorithmInterface.XmlEncrypmentMethodUrl());
+
+        //Encrypt the SymmetricAlgorithm key using the public key from a X509Certificate2.
+        X509CertificateImpl.InitializeX509Certificate(X509CertBase64Value, '', DotNetX509Certificate2);
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al:95]: Well, I actually changed it to _internal_ so it works, but is that not ok?

File: Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        EncryptedXmlImpl.Encrypt(XmlDocument, ElementToEncrypt, X509CertBase64Value, SymmetricAlgorithm);
+    end;
+
+    /// <summary>
+    /// Decrypts all EncryptedData elements of the XML document using the specified Assymetric Key.
+    /// </summary>
+    /// <param name="EncryptedDocument">The XmlDcoument to decrypt.</param>
+    /// <param name="EncryptionKey">The Asymmtric Key to use to decrypt the Symmetric Keys in the document.</param>
+    /// <returns>Returns true if decryption was successful, otherwise false.</returns>
+    procedure DecryptDocument(var EncryptedDocument: XmlDocument; EncryptionKey: Record "Signature Key"): Boolean
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al:47]: Should this perhaps be in the implementation Codeunits DecryptDocument instead?

File: Modules/System Tests/Cryptography Management/app.json [hunk 4]
```diff
@@ -68,6 +68,10 @@
                      {
                         "from":  132601,
                         "to":  132601
+                     },
+                     {
+                        "from":  132611,
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System Tests/Cryptography Management/app.json]: Fixed.

File: Modules/System Tests/Cryptography Management/src/EncryptedXmlTests.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+// ------------------------------------------------------------------------------------------------
+codeunit 132611 "EncryptedXml Tests"
+{
+    Subtype = Test;
+
+    var
+        EncryptedXml: Codeunit EncryptedXml;
+        LibraryAssert: Codeunit "Library Assert";
+
+    [Test]
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System Tests/Cryptography Management/src/EncryptedXmlTests.Codeunit.al]: This should be ok now.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        DotNetEncryptedXml.ReplaceElement(DotNetXmlElementToEncrypt, DotNetEncryptedData, false);
+
+        //Convert the encrypted DotNet XmlDocument to a XmlDocument.
+        XmlDotNetConvert.FromDotNet(DotNetXmlDocument, XmlDocument);
+    end;
+
+    procedure DecryptDocument(var EncryptedDocument: XmlDocument; EncryptionKey: Record "Signature Key"): Boolean
+    var
+        XmlDotNetConvert: Codeunit "Xml DotNet Convert";
+        SymmetricAlgorithmInterface: Interface SymmetricAlgorithm;
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al]: Removed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    procedure DecryptKey(EncryptedKey: XmlElement; EncryptionKey: Record "Signature Key"; UseOAEP: Boolean; var KeyBase64Value: Text): Boolean
+    var
+        X509CertificateImpl: Codeunit "X509Certificate2 Impl.";
+        XmlDocument: XmlDocument;
+        NamespaceManager: XmlNamespaceManager;
+        CipherValue: XmlNode;
+        DotNetEncryptedXml: DotNet EncryptedXml;
+        DotNetCipherBytes, DotNetKeyBytes : DotNet Array;
+        DotNetConvert: Dotnet Convert;
+        DotNetX509Certificate2: Dotnet X509Certificate2;
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al]: Removed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+        //Replace the encrypted data with the decrypted xml
+        DotNetEncryptedXml.ReplaceData(DotNetXmlElement, DecryptedData);
+
+        exit(true);
+    end;
+
+    procedure DecryptKey(EncryptedKey: XmlElement; EncryptionKey: Record "Signature Key"; UseOAEP: Boolean; var KeyBase64Value: Text): Boolean
+    var
+        X509CertificateImpl: Codeunit "X509Certificate2 Impl.";
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al]: Removed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    /// <param name="XmlDocument">The XmlDocument to encrypt.</param>
+    /// <param name="ElementToEncrypt">The name of the element to encrypt.</param>
+    /// <param name="X509CertBase64Value">The X509Certificate2 to use for the asymmetric encryption.</param>
+    procedure Encrypt(var XmlDocument: XmlDocument; ElementToEncrypt: Text; X509CertBase64Value: Text)
+    begin
+        EncryptedXmlImpl.Encrypt(XmlDocument, ElementToEncrypt, X509CertBase64Value);
+    end;
+
+    /// <summary>
+    /// Like the Encrypt(var XmlDocument, ElementToEncrypt, X509CertBase64Value) procedure but with
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al]: Fixed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+{
+    Access = Internal;
+
+    var
+        ElementNotFoundErr: Label 'The "%1" element was not found.', Comment = '%1: The name of the xml element that could not be found.';
+        XmlEncElementUrlTok: Label 'http://www.w3.org/2001/04/xmlenc#Element', Locked = true;
+        XmlEncRSA15UrlTok: Label 'http://www.w3.org/2001/04/xmlenc#rsa-1_5', Locked = true;
+        XmlEncUrlTok: Label 'http://www.w3.org/2001/04/xmlenc#', Locked = true;
+
+    procedure Encrypt(var XmlDocument: XmlDocument; ElementToEncrypt: Text; X509CertBase64Value: Text)
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al:18]: Fixed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+        //Encrypt the element
+        DotNetEncryptedData := DotNetEncryptedXml.Encrypt(DotNetXmlElementToEncrypt, DotNetX509Certificate2);
+        DotNetEncryptedXml.ReplaceElement(DotNetXmlElementToEncrypt, DotNetEncryptedData, false);
+
+        //Convert the encrypted DotNet XmlDocument to a XmlDocument.
+        XmlDotNetConvert.FromDotNet(DotNetXmlDocument, XmlDocument);
+    end;
+
+    procedure Encrypt(var XmlDocument: XmlDocument; ElementToEncrypt: Text; X509CertBase64Value: Text; SymmetricAlgorithm: Enum SymmetricAlgorithm)
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al:51]: Fixed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        DotNetEncryptedXml := DotNetEncryptedXml.EncryptedXml();
+        DecryptedData := DotNetEncryptedXml.DecryptData(DotNetEncryptedData, DotNetSymmetricAlgorithm);
+
+        //Replace the encrypted data with the decrypted xml
+        DotNetEncryptedXml.ReplaceData(DotNetXmlElement, DecryptedData);
+
+        exit(true);
+    end;
+
+    procedure DecryptKey(EncryptedKey: XmlElement; EncryptionKey: Record "Signature Key"; UseOAEP: Boolean; var KeyBase64Value: Text): Boolean
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al:204]: FIxed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+        foreach Ordinal in Enum::SymmetricAlgorithm.Ordinals() do begin
+            SymmetricAlgorithmInterface := Enum::SymmetricAlgorithm.FromInteger(Ordinal);
+            if SymmetricAlgorithmInterface.XmlEncrypmentMethodUrl() = DotNetEncryptedData.EncryptionMethod.KeyAlgorithm then begin
+                SymmetricAlgorithmInterface.GetInstance(DotNetSymmetricAlgorithm);
+                break;
+            end;
+        end;
+
+        if IsNull(DotNetSymmetricAlgorithm) then
+            Error('Unsupported symmetric algorithm.');
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al]: Fixed.

File: Modules/System Tests/Cryptography Management/src/EncryptedXmlTests.Codeunit.al [hunk 14]
```diff
@@ -0,0 +1,244 @@
+// ------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+// ------------------------------------------------------------------------------------------------
+codeunit 132611 "EncryptedXml Tests"
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System Tests/Cryptography Management/src/EncryptedXmlTests.Codeunit.al]: Fixed.

File: Modules/System Tests/Cryptography Management/src/EncryptedXmlTests.Codeunit.al [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+    Subtype = Test;
+
+    var
+        EncryptedXml: Codeunit EncryptedXml;
+        LibraryAssert: Codeunit "Library Assert";
+
+    [Test]
+    procedure EncryptXmlDocument()
+    var
+        EncryptionKey: Record "Signature Key";
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System Tests/Cryptography Management/src/EncryptedXmlTests.Codeunit.al]: I hope I have understood this correctly and that I found all places where it needs to be done.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+    var
+        ElementNotFoundErr: Label 'The "%1" element was not found.', Comment = '%1: The name of the xml element that could not be found.';
+        XmlEncElementUrlTok: Label 'http://www.w3.org/2001/04/xmlenc#Element', Locked = true;
+        XmlEncRSA15UrlTok: Label 'http://www.w3.org/2001/04/xmlenc#rsa-1_5', Locked = true;
+        XmlEncUrlTok: Label 'http://www.w3.org/2001/04/xmlenc#', Locked = true;
+
+    procedure Encrypt(var XmlDocument: XmlDocument; ElementToEncrypt: Text; X509CertBase64Value: Text)
+    var
+        XmlDotNetConvert: Codeunit "Xml DotNet Convert";
+        X509CertificateImpl: Codeunit "X509Certificate2 Impl.";
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXmlImpl.Codeunit.al]: Should be ok now.

File: Modules/System/Cryptography Management/src/AesCryptoServiceProviderImpl.Codeunit.al [hunk 17]
```diff
@@ -0,0 +1,25 @@
+codeunit 1467 "AesCryptoServiceProvider Impl." implements SymmetricAlgorithm
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/AesCryptoServiceProviderImpl.Codeunit.al:6]: Fixed.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+    /// <summary>
+    /// Decrypts an EncryptedKey xml element using an asymmetric algorithm.
+    /// </summary>
+    /// <param name="EncryptedKey">The EncryptedKey xml element with the key to be decrypted.</param>
+    /// <param name="EncryptionKey">The asymmetric key used to decrypt the symmetric key.</param>
+    /// <param name="UseOAEP">A value that specifies whether to use Optimal Asymmetric Encryption Padding (OAEP).</param>
+    /// <param name="KeyBase64Value">The Base64 encoded decrypted key value.</param>
+    /// <returns>Returns true if decryption was successful, otherwise false.</returns>
+    procedure DecryptKey(EncryptedKey: XmlElement; EncryptionKey: Record "Signature Key"; UseOAEP: Boolean; var KeyBase64Value: Text): Boolean
```
[the remark this answers is not in the document]
Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al:64]: Fixed.

File: Modules/System Tests/Cryptography Management/src/EncryptedXmlTest.Codeunit.al [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+    end;
+
+    [Test]
+    procedure EncryptXmlDocumentTripleDES()
+    var
+        EncryptionSignatureKey: Record "Signature Key";
+        XmlDocumentToEncrypt: XmlDocument;
+        NamespaceManager: XmlNamespaceManager;
+        EncryptedKey: XmlNode;
+        KeyBase64Data: Text;
```
Comment (PredragMaricic) [inline Modules/System Tests/Cryptography Management/src/EncryptedXmlTest.Codeunit.al:52]: unused var

Comment (kennetlindberg) [reply Modules/System Tests/Cryptography Management/src/EncryptedXmlTest.Codeunit.al]: Fixed

**Comment by PredragMaricic** [comment]
I think we are ready for Microsoft internal review and submission. Will keep you updated, but hopefully it should reach the 19.1. minor update.

File: Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+    /// <summary>
+    /// Creates a symmetric session key to encrypt a XML document and then 
+    /// uses the X.509 certificate to embed an encrypted version of the session key in the XML document.
+    /// </summary>
+    /// <param name="XmlDocument">The XmlDocument to encrypt.</param>
+    /// <param name="ElementToEncrypt">The name of the element to encrypt.</param>
+    /// <param name="X509CertBase64Value">The X509Certificate2 to use for the asymmetric encryption.</param>
+    [NonDebuggable]
+    procedure Encrypt(var XmlDocument: XmlDocument; ElementToEncrypt: Text; X509CertBase64Value: Text)
```
Comment (PredragMaricic) [inline Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al:24]: Can you please add support for certs with passwords ? We have most of the cases where certs have pass.

Comment (kennetlindberg) [reply Modules/System/Cryptography Management/src/Xml/EncryptedXml.Codeunit.al]: Yes, of course, that is better. Fixed now.

**Comment by kennetlindberg** [comment]
Is it possible to say when this will be merged into master, will it be available in 19.1? Is there anything I can do to help?

**Comment by PredragMaricic** [comment]
> Is it possible to say when this will be merged into master, will it be available in 19.1? Is there anything I can do to help?

Thanks @kennetlindberg. It passed the internal code review now. I will try to make it ready for 19.1, but cannot promise. Worse case it will go into 19.2.

**Comment by PredragMaricic** [comment]
Hey @kennetlindberg,

We had a bug not related to your PR, that affected the interface. Can you please send me your email, so that we can discuss those changes?

Thanks,
Predrag

**Comment by JesperSchulz** [comment]
The code was merged. Will repoen and merge this PR.

**Comment by JesperSchulz** [comment]
Availability update: We will publish a fix for this issue in the next update for release 19.

**Pull request merged**